design/transaction-layer-refactoring.md

```diff
@@ -0,0 +1,151 @@
+    for (mutation, guard) in req.mutations.into_iter().zip(&guards) {
+        match txn.acquire_pessimistic_lock(...) {
+            Ok(...) => ...,
+            Err(KeyIsLocked(lock_info)) => {
+                guard.set_lock(lock_info.clone());
+                let lock_released = guard.lock_released(); // returns a future which is ready when the lock is released
+                locked = Some(lock_info, );
+            },
+            Err(e) => ...
+        }
+    }
+    
+    if let Some((lock_info, lock_released)) = locked {
+        drop(guards); // release the latches first
+        lock_released.await; // easy to add timeout
+        resp.set_errors(...);
+    } else {
+        if self.cfg.pipelined_pessimistic_lock {
+            // write to the raft store asynchronously
+            let engine = self.engine.clone();
+            self.spawn(async move {
+                engine.write(txn.into_modifies()).await;
+                drop(guards);
+            });
+        } else {
+            // write to the raft store synchronously
+            self.engine.write(txn.into_modifies()).await; 
+        }
+        ...
+    }
+    resp
+}
+```
+
+The goal is to put the whole process of each command inside a single function. Then, people only need to look at one function to learn the process. Code related to transactions will be more understandable.
+
+Moreover, long code paths and jumps are avoided. It's never a problem that dependencies and configurations need be passed through the long path.
+
+## In-memory lock table
+
+Both the latch and the lock manager stores memory locks and notify when the locks are released. For "parallel commit", we also need another memory locking mechanism. It'll be good to have an integrated locking mechanism handling all these requirements.
+
+We can use a concurrent ordered map to build a lock table. We map each raw key to a memory lock. The memory lock contains lock information and waiting lists. Currently we have two kinds of orthogonal waiting list: the latch waiting list and the pessimistic lock waiting list. 
+
+```rust
+pub type LockTable = ConcurrentOrderedMap<Vec<u8>, Arc<MemoryLock>>;
+
+pub struct MemoryLock {
+    mutex_state: AtomicU64,
+    mutex_waiters: ConcurrentList<Notify>
+    lock_info: Mutex<Option<LockInfo>>,
+    pessimistic_waiters: ConcurrentList<Notify>,
+}
+```
+
+Both the original latches and lock manager can be implemented with this memory lock.
+
+For the original latch usage, the lock serves as an asynchronous mutex. It can return a future that outputs a guard. The guard can be used to modify the data in the memory lock. When the guard is dropped, other tasks waiting for the lock are notified.
+
+For the lock manager usage, it provides the functionality to add waiters and modify the lock information. When `AcquirePessimisticLock` meets a lock, it adds itself to the waiting list and stores the lock information. When the lock is cleared, the waiters are all notified.

```

Comment:
Is this flexible enough if we want to manage the waiting lists? E.g., timeout if waiting too long, or prioritise some requests over others

Reply:
Timeout can be implemented using composition of futures. It needn't be integrated to the lock table.
When we register to the waiting list, we may add extra information to it like priority or a specific timestamp. It can be powerful if we want, but it can be complex...

Reply:
That's ok, as long as there is some avenue for extension in the future and we're not constraining our future selves